### PR TITLE
Document C string literal tokens.

### PR DESCRIPTION
**Pull request by jmillikin**

### CODE REVIEW

**Comment by jmillikin** [comment]
Note: this feature is being stabilized in https://github.com/rust-lang/rust/pull/117472 -- CI will fail until run with ` rustc` that includes that PR.

I ran `mdbook test && mdbook build` locally to verify that the tests pass when run with a `rustdoc` containing the stabilization PR.

File: src/tokens.md [hunk 1]
```diff
@@ -328,6 +330,76 @@ b"\x52"; b"R"; br"R";                // R
...
+### C string and raw C string literals
+
+#### C string literals
+
+> **<sup>Lexer</sup>**\
+> C_STRING_LITERAL :\
+> &nbsp;&nbsp; `c"` ( ASCII_FOR_C_STRING | BYTE_ESCAPE | STRING_CONTINUE )<sup>\*</sup> `"` SUFFIX<sup>?</sup>
+>
+> ASCII_FOR_C_STRING :\
+> &nbsp;&nbsp; _any non-NUL ASCII (i.e 0x01 to 0x7F), except_ `"`, `\` _and IsolatedCR_
```
Comment (ehuss) [inline src/tokens.md:342]: This doesn't seem quite correct to me, since it looks like the parser allows a bare NUL in the string (there are later stages that perform the validation). The grammar specified in these sections needs to specify the format that the parser accepts. The prose below should explain any further restrictions placed on it.

Additionally, it looks like it does accept utf-8 input, so I'm not sure why it mentions the ascii restriction.

File: src/tokens.md [hunk 2]
```diff
@@ -328,6 +330,76 @@ b"\x52"; b"R"; br"R";                // R
...
+Alternatively, a C string literal can be a _raw C string literal_, defined
+below. The type of a C string literal is `&core::ffi::CStr`.
+
+Some additional _escapes_ are available in either C or non-raw C string
+literals. An escape starts with a `U+005C` (`\`) and continues with one of the
+following forms:
+
+* A _byte escape_ escape starts with `U+0078` (`x`) and is followed by exactly
+  two _hex digits_. It denotes the byte equal to the provided hex value. The
+  byte escape sequence `\x00` is forbidden, as C strings may not contain `NUL`.
```
Comment (ehuss) [inline src/tokens.md:357]: I might reword this to cover both a bare NUL and `\x00` are not allowed.

File: src/tokens.md [hunk 3]
```diff
@@ -328,6 +330,76 @@ b"\x52"; b"R"; br"R";                // R
...
+  escaped in order to denote its ASCII encoding `0x5C`.
+
+#### Raw C string literals
+
+> **<sup>Lexer</sup>**\
+> RAW_C_STRING_LITERAL :\
+> &nbsp;&nbsp; `cr` RAW_C_STRING_CONTENT SUFFIX<sup>?</sup>
+>
+> RAW_C_STRING_CONTENT :\
+> &nbsp;&nbsp; &nbsp;&nbsp; `"` ASCII_EXCEPT_NUL<sup>* (non-greedy)</sup> `"`\
```
Comment (ehuss) [inline src/tokens.md:371]: Similar comments as above about what is accepted here.

File: src/tokens.md [hunk 4]
```diff
@@ -328,6 +330,76 @@ b"\x52"; b"R"; br"R";                // R
...
+> &nbsp;&nbsp; | `#` RAW_C_STRING_CONTENT `#`
+>
+> ASCII_EXCEPT_NUL :\
+> &nbsp;&nbsp; _any non-NUL ASCII (i.e. 0x01 to 0x7F)_
+
+Raw C string literals do not process any escapes. They start with the
+character `U+0063` (`c`), followed by `U+0072` (`r`), followed by fewer than 256
+of the character `U+0023` (`#`), and a `U+0022` (double-quote) character. The
+_raw string body_ can contain any sequence of non-`NUL` ASCII characters and is terminated
```
Comment (ehuss) [inline src/tokens.md:380]:
```suggestion
_raw C string body_ can contain any sequence of non-`NUL` ASCII characters and is terminated
```

Also similar comments about being ascii and having NUL.

File: src/tokens.md [hunk 5]
```diff
@@ -328,6 +330,76 @@ b"\x52"; b"R"; br"R";                // R
 b"\\x52"; br"\x52";                  // \x52
 ```
 
+### C string and raw C string literals
+
+#### C string literals
+
+> **<sup>Lexer</sup>**\
+> C_STRING_LITERAL :\
+> &nbsp;&nbsp; `c"` ( ASCII_FOR_C_STRING | BYTE_ESCAPE | STRING_CONTINUE )<sup>\*</sup> `"` SUFFIX<sup>?</sup>
```
Comment (ehuss) [inline src/tokens.md:339]: This seems to be missing unicode escapes?

**Review by ehuss — COMMENTED** [review]
Thanks!

Can you also include a section that indicates that C-string literals are only available in Edition 2021 or newer? Edition differences are specified in blockquotes (search for "Edition Differences" for the format).

I believe the [Reserved prefixes](https://github.com/rust-lang/reference/blob/HEAD/src/tokens.md#reserved-prefixes) section needs to be updated with `c` and `cr` being excluded.

I believe [Literal patterns](https://github.com/rust-lang/reference/blob/HEAD/src/patterns.md#literal-patterns) will need to be updated since C-strings are accepted there syntactically. (They can't really be used since CStr doesn't implement Eq/PartialEq, though.)

File: src/tokens.md [hunk 6]
```diff
@@ -328,6 +330,76 @@ b"\x52"; b"R"; br"R";                // R
...
+>
+> ASCII_FOR_C_STRING :\
+> &nbsp;&nbsp; _any non-NUL ASCII (i.e 0x01 to 0x7F), except_ `"`, `\` _and IsolatedCR_
+
+A non-raw _C string literal_ is a sequence of ASCII characters and _escapes_,
+preceded by the characters `U+0063` (`c`) and `U+0022` (double-quote), and
+followed by the character `U+0022`. If the character `U+0022` is present within
+the literal, it must be _escaped_ by a preceding `U+005C` (`\`) character.
+Alternatively, a C string literal can be a _raw C string literal_, defined
+below. The type of a C string literal is `&core::ffi::CStr`.
```
[the remark this answers is not in the document]
Comment (jmillikin) [reply src/tokens.md]: Done (added link + mentioned `NUL` terminator)

File: src/tokens.md (hunk 1 above)
Comment (jmillikin) [reply src/tokens.md]: Ah, interesting, I didn't realize the reference distinguished the grammar and permitted literal content. Fixed.

The ASCII restriction was just me misreading the RFC -- I didn't realize Unicode was permitted. I've fixed the notation so that it describes C string literals as accepting both byte escapes (like `b""`) and Unicode escapes (like `""`).

File: src/tokens.md (hunk 2 above)
Comment (jmillikin) [reply src/tokens.md]: Done-ish -- I moved the section on `NUL` escapes to the bottom of this section, to make it clearer that `\0`, `\x00`, and `\u{0000}` are treated uniformly.

File: src/tokens.md (hunk 3 above)
Comment (jmillikin) [reply src/tokens.md]: Done.

File: src/tokens.md (hunk 4 above)
Comment (jmillikin) [reply src/tokens.md]: Done.

File: src/tokens.md (hunk 5 above)
Comment (jmillikin) [reply src/tokens.md]: Fixed, thanks! (see earlier comments regarding Unicode misunderstanding)

**Comment by jmillikin** [comment]
Thank you for the quick review! I've made the recommended changes, fixed the ASCII vs Unicode misunderstanding, and tried to clarify the wording around `NUL` escapes.

diff: https://github.com/rust-lang/reference/compare/5d1950799abab6174d8a797952889851dbe2774b..2481014bf2a856425e2c0d9971b8f77becb6eebc

File: src/tokens.md [hunk 7]
```diff
@@ -328,6 +330,95 @@ b"\x52"; b"R"; br"R";                // R
...
+> &nbsp;&nbsp; &nbsp;&nbsp; | UNICODE_ESCAPE\
+> &nbsp;&nbsp; &nbsp;&nbsp; | STRING_CONTINUE\
+> &nbsp;&nbsp; )<sup>\*</sup> `"` SUFFIX<sup>?</sup>
+
+A _C string literal_ is a sequence of Unicode characters and _escapes_,
+preceded by the characters `U+0063` (`c`) and `U+0022` (double-quote), and
+followed by the character `U+0022`. If the character `U+0022` is present within
+the literal, it must be _escaped_ by a preceding `U+005C` (`\`) character.
+Alternatively, a C string literal can be a _raw C string literal_, defined
+below. The type of a C string literal is [`&core::ffi::CStr`][CStr].
```
Comment (eggyal) [inline src/tokens.md:351]: Whilst below it is mentioned that code point escapes are encoded as UTF-8, nowhere is it stated how the Unicode characters contained within the C string literal are encoded in the ensuing `CStr`: I presume also UTF-8?  Perhaps this should be stated explicitly for the avoidance of any doubt.

**Review by eggyal — CHANGES_REQUESTED** [review]

File: src/tokens.md (hunk 7 above)
Comment (jmillikin) [reply src/tokens.md:351]: Done -- I added a section about encoding after the escapes.

**Comment by jmillikin** [comment]
The stabilization PR has merged and this PR's CI build is now green.
